Shell autocomplete functions do not work correctly

<img src="https://raw.githubusercontent.com/derailed/k9s/master/assets/k9s_err.png" align="right" width="100" height="auto"/>

<br/>
<br/>
<br/>

**Describe the bug**
k9s includes shell autocompletion commands that appear to have been taken directly from kubectl except they don't work as expected.

I work with a number of contexts and switching between them, especially with OIDC, is quite slow. I prefer to start k9s with a `--context` flag to select my desired context upon startup but the lack of completion suggestions is frustrating.

The undocumented `__complete` autocomplete suggestions command only works to suggest the top level functions but does not work for suggesting options to double-dash parameters like `--context`.

**To Reproduce**
```
% k9s __complete ""
completion	Generate the autocompletion script for the specified shell
help	Help about any command
info	List K9s configurations info
version	Print version/build info
:4
Completion ended with directive: ShellCompDirectiveNoFileComp

% k9s __complete - ""
:0
Completion ended with directive: ShellCompDirectiveDefault

% k9s __complete --context ""
:0
Completion ended with directive: ShellCompDirectiveDefault
```

~Furthermore the command suggests that `--completion`, `--help`, `--info`, `--version` should also work but they end in a panic except for `--help`.~ [EDIT: - this is an incorrect statement but k9s should handle incorrect flags without panicking]
```
% k9s __complete -- ""
completion	Generate the autocompletion script for the specified shell
help	Help about any command
info	List K9s configurations info
version	Print version/build info
:4
Completion ended with directive: ShellCompDirectiveNoFileComp

% k9s --version
Error: unknown flag: --version
[-snipped-]
panic: unknown flag: --version

goroutine 1 [running]:
github.com/derailed/k9s/cmd.Execute(...)
	github.com/derailed/k9s/cmd/root.go:60
main.main()
	github.com/derailed/k9s/main.go:32 +0x3c
```

**Expected behavior**
k9s should be able to suggest completion options for flags the same way kubectl does, eg.
```
% kubectl __complete --user ""
local
oidc
:4
Completion ended with directive: ShellCompDirectiveNoFileComp

% k9s __complete --user ""
:0
Completion ended with directive: ShellCompDirectiveDefault
```
and
```
% kubectl __complete --context ""
microk8s
k3s
:4
Completion ended with directive: ShellCompDirectiveNoFileComp

% k9s __complete --context ""
:0
Completion ended with directive: ShellCompDirectiveDefault
```


**Versions:**

Seems to affect all versions
```
% k9s version
 ____  __.________
|    |/ _/   __   \______
|      < \____    /  ___/
|    |  \   /    /\___ \
|____|__ \ /____//____  >
        \/            \/

Version:    v0.31.5
Commit:     ff17fff2861c02d325fc95d536d3787b83cb3782
Date:       2024-01-12T22:03:40Z
```

**Additional context**
It took me quite a while to work out that the output from `k9s completion zsh` actually calls `k9s __complete <flags> "string"` to get its suggestions.

--- a/cmd/root.go
+++ b/cmd/root.go
@@ -4,11 +4,14 @@
 package cmd
 
 import (
+	"errors"
 	"fmt"
 	"os"
 	"runtime/debug"
+	"strings"
 
 	"github.com/derailed/k9s/internal/config/data"
+	"k8s.io/client-go/tools/clientcmd/api"
 
 	"github.com/derailed/k9s/internal/client"
 	"github.com/derailed/k9s/internal/color"
@@ -44,10 +47,18 @@ var (
 	out = colorable.NewColorableStdout()
 )
 
+type FlagError struct{ err error }
+
+func (e *FlagError) Error() string { return e.err.Error() }
+
 func init() {
 	if err := config.InitLogLoc(); err != nil {
 		fmt.Printf("Fail to init k9s logs location %s\n", err)
 	}
+
+	rootCmd.SetFlagErrorFunc(func(command *cobra.Command, err error) error {
+		return &FlagError{err: err}
+	})
 
 	rootCmd.AddCommand(versionCmd(), infoCmd())
 	initK9sFlags()
@@ -57,7 +68,10 @@ func init() {
 // Execute root command.
 func Execute() {
 	if err := rootCmd.Execute(); err != nil {
-		panic(err)
+		var flagError *FlagError
+		if !errors.As(err, &flagError) {
+			panic(err)
+		}
 	}
 }
 
@@ -281,6 +295,7 @@ func initK8sFlags() {
 
 	initAsFlags()
 	initCertFlags()
+	initK8sFlagCompletion()
 }
 
 func initAsFlags() {
@@ -334,4 +349,52 @@ func initCertFlags() {
 		"",
 		"Bearer token for authentication to the API server",
 	)
+}
+
+func initK8sFlagCompletion() {
+	_ = rootCmd.RegisterFlagCompletionFunc("context", k8sFlagCompletionFunc(func(cfg *api.Config) map[string]*api.Context {
+		return cfg.Contexts
+	}))
+
+	_ = rootCmd.RegisterFlagCompletionFunc("cluster", k8sFlagCompletionFunc(func(cfg *api.Config) map[string]*api.Cluster {
+		return cfg.Clusters
+	}))
+
+	_ = rootCmd.RegisterFlagCompletionFunc("user", k8sFlagCompletionFunc(func(cfg *api.Config) map[string]*api.AuthInfo {
+		return cfg.AuthInfos
+	}))
+
+	_ = rootCmd.RegisterFlagCompletionFunc("namespace", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		conn, err := client.InitConnection(client.NewConfig(k8sFlags))
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+
+		nss, err := conn.ValidNamespaceNames()
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+
+		return filterFlagCompletions(nss, toComplete)
+	})
+}
+
+func k8sFlagCompletionFunc[T any](picker func(cfg *api.Config) map[string]T) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		k8sCfg, err := client.NewConfig(k8sFlags).RawConfig()
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+		return filterFlagCompletions(picker(&k8sCfg), toComplete)
+	}
+}
+
+func filterFlagCompletions[T any](m map[string]T, toComplete string) ([]string, cobra.ShellCompDirective) {
+	var completions []string
+	for name := range m {
+		if strings.HasPrefix(name, toComplete) {
+			completions = append(completions, name)
+		}
+	}
+	return completions, cobra.ShellCompDirectiveNoFileComp
 }

--- a/internal/client/client.go
+++ b/internal/client/client.go
@@ -71,7 +71,7 @@ func InitConnection(config *Config) (*APIClient, error) {
 	if err != nil {
 		log.Error().Err(err).Msgf("Fail to locate metrics-server")
 	}
-	if errors.Is(err, noMetricServerErr) || errors.Is(err, metricsUnsupportedErr) {
+	if err == nil || errors.Is(err, noMetricServerErr) || errors.Is(err, metricsUnsupportedErr) {
 		return &a, nil
 	}
 	a.connOK = false

